veecle-osal-embassy/src/thread.rs

```diff
@@ -0,0 +1,95 @@
+//! Thread-related abstractions.
+
+#[cfg(not(target_os = "none"))]
+use std::cell::Cell;
+#[cfg(not(target_os = "none"))]
+use std::sync::atomic::{AtomicU64, Ordering};
+pub use veecle_osal_api::thread::ThreadAbstraction;
+
+/// Implements the [`ThreadAbstraction`] trait for Embassy.
+///
+/// Only supports running on a single core (and thread) on `no_std` systems.
+/// Using the abstraction on any `no_std` system that uses multiple cores or threads may lead to undefined behavior.
+#[derive(Debug)]
+pub struct Thread;
+
+/// Global counter for generating unique thread ids.
+#[cfg(not(target_os = "none"))]
+static NEXT_THREAD_ID: AtomicU64 = AtomicU64::new(1);
+
+#[cfg(not(target_os = "none"))]
+std::thread_local! {
+    /// Thread-local storage for the current thread's id.
+    static THREAD_ID: Cell<u64> = const { Cell::new(0) };
+}
+
+impl ThreadAbstraction for Thread {
+    #[cfg(not(target_os = "none"))]
+    fn current_thread_id() -> u64 {
+        THREAD_ID.with(|id| {
+            let current = id.get();
+            if current == 0 {
+                // `Relaxed` here is enough because we don't care about what values various threads
+                // see, just that they're unique (assuming that creating 2^64 threads is impractical
+                // so overflow can't happen).
+                let new_id = NEXT_THREAD_ID.fetch_add(1, Ordering::Relaxed);
+                id.set(new_id);
+                new_id
+            } else {
+                current
+            }
+        })
+    }
+
+    #[cfg(target_os = "none")]
+    fn current_thread_id() -> u64 {
+        1
+    }
+}
```

Comment:
It could be more readable to have inline submodules for the different `cfg` since they don't share code, avoiding the proliferation of `cfg`s on many items

```rust
#[cfg(not(target_os = "none"))]
mod os {
  use std::cell::Cell;
  ...
  impl ThreadAbstraction for Thread {
    #[cfg(not(target_os = "none"))]
    fn current_thread_id() -> u64 {
      ...
    }
  }
}

#[cfg(target_os = "none")]
mod none {
  impl ThreadAbstraction for Thread {
    #[cfg(target_os = "none")]
    fn current_thread_id() -> u64 {
        1
    }
  }
}
```

(Or in this case the same could be achieved by just putting all the top-level cfg-ed items inside the method)